admission_control/admission-control-service/src/upstream_proxy.rs

```diff
@@ -6,198 +6,234 @@ use admission_control_proto::proto::admission_control::{
-        }
+/// new peer discovery handler
+/// adds new entry to `peer_info`
+fn new_peer(peer_info: &mut HashMap<PeerId, bool>, peer_id: PeerId) {
+    if let Some(state) = peer_info.get_mut(&peer_id) {
+        *state = true;
     }
+}
 
-    /// lost peer handler. Marks connection as dead
-    fn lost_peer(&mut self, peer_id: PeerId) {
-        if let Some(state) = self.peer_info.get_mut(&peer_id) {
-            *state = false;
-        }
+/// lost peer handler. Marks connection as dead
+fn lost_peer(peer_info: &mut HashMap<PeerId, bool>, peer_id: PeerId) {
+    if let Some(state) = peer_info.get_mut(&peer_id) {
+        *state = false;
     }
+}
 
-    async fn submit_transaction_upstream(
-        &mut self,
-        request: SubmitTransactionRequest,
-    ) -> failure::Result<SubmitTransactionResponse> {
-        let active_peer_ids = self.get_active_upstream_peers();
-        if !active_peer_ids.is_empty() {
-            let peer_id = { active_peer_ids.choose(&mut rand::thread_rng()) };
-            if let Some(peer_id) = peer_id {
-                let result = self
-                    .network_sender
-                    .send_transaction_upstream(
-                        peer_id.clone(),
-                        request,
-                        self.ac_config.upstream_proxy_timeout,
-                    )
-                    .await?;
-                return Ok(result);
-            }
-        }
-        Err(format_err!("[admission-control] No active upstream peers"))
+fn pick_peer(peer_info: &HashMap<PeerId, bool>) -> Option<PeerId> {
+    let active_peer_ids: Vec<PeerId> = peer_info
+        .iter()
+        .filter(|(_, &is_alive)| is_alive)
+        .map(|(&peer_id, _)| peer_id)
+        .collect();
+    if !active_peer_ids.is_empty() {
+        return active_peer_ids.choose(&mut rand::thread_rng()).cloned();
     }
+    None
+}
 
-    async fn process_submit_transaction_request(
-        &mut self,
-        request: SubmitTransactionRequest,
-        callback: oneshot::Sender<Result<Bytes, RpcError>>,
-    ) -> failure::Result<()> {
-        let mut response_msg = None;
-        match self.role {
-            RoleType::Validator => {
-                let result = self.client.submit_transaction(&request);
-                match result {
-                    Ok(response) => {
-                        let ac_control_msg = AdmissionControlMsg {
-                            message: Some(AdmissionControlMsg_oneof::SubmitTransactionResponse(
-                                response,
-                            )),
-                        };
-                        response_msg = Some(ac_control_msg);
-                    }
-                    Err(e) => {
-                        return Err(format_err!("{:?}", e.to_string()));
-                    }
-                }
+async fn start_submit_transaction_upstream(
+    request: SubmitTransactionRequest,
+    upstream_proxy_data: UpstreamProxyData,
+    peer_id: Option<PeerId>,
```

Comment:
nit: probably would be cleaner to have api here that accepts PeerId as parameter. Not Option. And filter out None case on callsite side